tests/compiletest.rs

```diff
@@ -1,71 +1,432 @@
+        }
     }
+    flags.push("-Zui-testing".to_string());
+    flags.push("--target".to_string());
+    flags.push(target.to_string());
+    let target = target.to_string();
 
-    let flags = flags.join(" ");
-    eprintln!("   Compiler flags: {}", flags);
+    eprintln!("   Compiler flags: {:?}", flags);
 
-    // The rest of the configuration.
-    let mut config = compiletest::Config::default().tempdir();
-    config.mode = mode.parse().expect("Invalid mode");
-    config.rustc_path = miri_path();
-    if let Some(lib_path) = option_env!("RUSTC_LIB_PATH") {
-        config.run_lib_path = PathBuf::from(lib_path);
-        config.compile_lib_path = PathBuf::from(lib_path);
-    }
-    config.filters = env::args().nth(1).into_iter().collect();
-    config.host = get_host();
-    config.src_base = PathBuf::from(path);
-    config.target = target.to_owned();
-    config.target_rustcflags = Some(flags);
-    compiletest::run_tests(&config);
-}
+    let grab_entries =
+        |path: &Path| std::fs::read_dir(path).unwrap().map(|entry| entry.unwrap().path());
+    let todo = SegQueue::new();
+    todo.push(PathBuf::from(path));
+
+    let failures = Mutex::new(vec![]);
+    let total = AtomicUsize::default();
+    let skipped = AtomicUsize::default();
+
+    crossbeam::scope(|s| {
+        for _ in 0..num_cpus::get() {
+            s.spawn(|_| {
+                while let Some(path) = todo.pop() {
+                    // Collect everything inside directories
+                    if path.is_dir() {
+                        for entry in grab_entries(&path) {
+                            todo.push(entry);
+                        }
+                        continue;
+                    }
+                    // Only look at .rs files
+                    if let Some(ext) = path.extension() {
+                        if ext != "rs" {
+                            continue;
+                        }
+                    } else {
+                        continue;
+                    }
+                    total.fetch_add(1, Ordering::Relaxed); // Read rules for skipping from file
+                    if ignore_file(&path, &target) {
+                        skipped.fetch_add(1, Ordering::Relaxed);
+                        eprintln!("{} .. {}", path.display(), "skipped".yellow());
+                        continue;
+                    }
+                    for revision in revisions(&path) {
+                        let (m, errors) = run_test(&path, &target, &flags, mode, &revision);
 
-fn compile_fail(path: &str, target: &str) {
+                        eprint!("{} .. ", path.display());
+                        if errors.is_empty() {
+                            eprintln!("{}", "ok".green());
+                        } else {
+                            eprint!("{}", "FAILED".red().bold());
+                            if !revision.is_empty() {
+                                eprint!(" (revision `{}`)", revision);
```

Comment:
Why do we show the revision only if it failed? And after the state?
I would have expected something more like `{path.display()} (revision: {revision}) .. {status}`